File: leetcode/62_Unique_Paths.py
import unittest

class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [1] * n
        for _ in range(1, m):
            for i in range(1,n):
                dp[i] += dp[i - 1]
        return dp[n - 1]

class TestUniquePaths(unittest.TestCase):
    def test_1(self):
        self.assertEqual(Solution().uniquePaths(3, 7), 28)

    def test_2(self):
        self.assertEqual(Solution().uniquePaths(3, 2), 3)

    def test_3(self):
        self.assertEqual(Solution().uniquePaths(7, 3), 28)

    def test_4(self):
        self.assertEqual(Solution().uniquePaths(3, 3), 6)

File: leetcode/62_Unique_Paths.rs
struct Solution {}

impl Solution {
    pub fn unique_paths(m: i32, n: i32) -> i32 {
        let (n, m) = (n as usize, m as usize);
        let mut dp = vec![1; n];
        for _ in 1..m {
            for i in 1..n {
                dp[i] += dp[i - 1];
            }
        }       
        dp[n - 1] 
    }
}

#[cfg(test)]
mod unique_paths_test {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(Solution::unique_paths(3, 7), 28);
    }    

    #[test]
    fn test_2() {
        assert_eq!(Solution::unique_paths(3, 2), 3);
    }    

    #[test]
    fn test_3() {
        assert_eq!(Solution::unique_paths(7, 3), 28);
    }    

    #[test]
    fn test_4() {
        assert_eq!(Solution::unique_paths(3, 3), 6);
    }    
}
